### ISSUE TITLE
copy and paste do not work on ipad

### ISSUE THREAD

**Issue by williamstein** [xtermjs/xterm.js#3727]
## Details
- Browser and browser version: Safari on an iPad
- OS version: all versions of iPadOS (and iPhoneOS too)
- xterm.js version: all versions

### Steps to reproduce

1. Go to the demo at https://xtermjs.org
2. The terminal works. However, it is not possible to select text at all by touching the screen, so copy (and paste) are impossible.
3. Using a keyboard and trackpad on the iPad, you can select text, but when you copy with (cmd+c), it does NOT copy, leaving the clipboard unchanged.  Paste does work with a keyboard.
4. I also tried both the canvas and dom, and the webgl addon.  I couldn't select text with any of those.

I use xtermjs, in https://cocalc.com for desktop, but for mobile I've been using an ancient version of term.js... almost entirely because copy paste actually always fully worked with term.js on ipad.

**Comment by williamstein** [comment]
Minor correction: paste never worked in the ancient term.js; however, copy does.

**Comment by Tyriar** [comment]
Copy and paste may work with the DOM renderer (and/or screen reader mode) on iPad, pretty sure this isn't possible with canvas/webgl until we get better APIs from the browser though.

**Comment by williamstein** [comment]
> Copy and paste may work with the DOM renderer (and/or screen reader mode) on iPad

Thanks, but unfortunately, it doesn't yet.  As it used to with the term.js DOM renderer, I'm optimistic it's possible.  Honestly, I'm very surprised it doesn't work already, and hope it's just a matter of some small CSS/Javascript tweak to fix this.   I'm definitely fine with DOM-only for iPad/iOS.       I'll likely dive into fixing this bug myself and contributing the fix, as soon as I get time.

**Comment by andy380743909** [comment]
The latest version 5.3.0, I set the rendererType to 'dom' with this code `term.options.rendererType = 'dom';`, but the selection function still not work on my iPhone.

**Comment by williamstein** [comment]
> The latest version 5.3.0, I set the rendererType to 'dom' with this code term.options.rendererType = 'dom';, but the selection function still not work on my iPhone.

Thanks for testing.  I'm still stuck shipping the ancient term.js that xterm.js was based on, because of this.  I've never been able to figure out how to get this to work on mobile with dom rendering either.  It can't be hard... but how?

**Comment by jerch** [comment]
> It can't be hard...

Yeah, it prolly isn't (although the event model on a touch device differs alot).

Our whole testing is very desktop-centered, so for mobile devices we def. have a blind spot. Not only for iOS devices, we also see a lot of issues on Android around key input and touch navigation.

Unless a white knight with strong touch device dedication shows up, this is unlikely to change soon. So any help in that field is more than welcome.

**Comment by williamstein** [comment]
Thanks for the encouragement; I will try again when I get a chance.  Please don't close this issue as wontfix, since maybe I'll fix it.  It definitely matters to me a lot, and I would love to not also ship term.js.    I definitely don't want to solve all the touch navigation and input issues -- I just want to make copy possible.  On cocalc I have other solutions to navigation issues, e.g., there are buttons across the top for standard things like tab that you need in a terminal, and there is also a normal text area that one type into instead at the top, and then submit the input.  But not being able to copy from the rendered output is a real dealbreaker in terms of usability.

**Comment by ValerioB88** [comment]
I confirm it doesn't work on android either (tested on the xterm on xtermjs.org). Would love to have this fixed :)

**Comment by friuns2** [comment]
Not working:(

**Comment by Tyriar** [comment]
Please don't comment something like +1 or "me too" as it sends out notifications to everyone. Vote on the top comment for that:

<img width="1060" height="214" alt="Image" src="https://github.com/user-attachments/assets/2841d9fe-4992-411b-a7f6-3b6154878567" />

**Comment by LeeHolmes** [comment]
xterm has a lot of code and styles that interfere with selection so that it can get that cool green selection view.

I've implemented a workaround that works well, although is truly a workaround: https://www.leeholmes.com/projects/[pyco](https://www.leeholmes.com/projects/pyco/)/

The majority of it came in with this change: https://github.com/LeeHolmes/pycoterm/commit/378c0c2654141adfd2db98e6f62c3672058d3ac1, but you can look at headrev for the current full implementation: https://github.com/LeeHolmes/pycoterm/blob/main/pyco_web/index.html

That change also has a bunch of code to allow terminal scrolling via traditional gestures vs. the itty bitty scroll bar.

In brief:

- Capture long press
- When that happens, display a text overlay over top xterm that's populated from the syntax-highlighted content of the terminal
- When selection goes away, hide the overlay again

I also explored capturing the long press and then updating the xterm styles that were preventing traditional text selection and then putting them back when done. That worked as well, but was both more code and far more fragile.

**Comment by williamstein** [comment]
Hopefully an actual fix here (not a workaround): https://github.com/xtermjs/xterm.js/pull/5961

Package @cocalc/xterm in case you want to easily try this out: https://www.npmjs.com/package/@cocalc/xterm
